File: src/auth/jwt.py
"""JWT token creation and validation."""

from datetime import datetime, timedelta
from typing import Optional

from jose import JWTError, jwt

from src.config.settings import settings


class TokenError(Exception):
    """Exception raised for token-related errors."""

    pass


def create_access_token(
    user_id: str,
    expires_delta: Optional[timedelta] = None,
) -> str:
    """Create a new access token.

    Args:
        user_id: User UUID to encode in the token.
        expires_delta: Optional custom expiration time.

    Returns:
        Encoded JWT access token.
    """
    if expires_delta is None:
        expires_delta = timedelta(minutes=settings.access_token_expire_minutes)

    expire = datetime.utcnow() + expires_delta
    to_encode = {
        "sub": user_id,
        "exp": expire,
        "type": "access",
    }
    return jwt.encode(
        to_encode,
        settings.jwt_secret_key,
        algorithm=settings.jwt_algorithm,
    )


def create_refresh_token(
    user_id: str,
    remember_me: bool = False,
) -> tuple[str, datetime]:
    """Create a new refresh token.

    Args:
        user_id: User UUID to encode in the token.
        remember_me: Whether to use extended expiration.

    Returns:
        Tuple of (encoded JWT refresh token, expiration datetime).
    """
    if remember_me:
        expires_delta = timedelta(days=settings.refresh_token_expire_days_remember)
    else:
        expires_delta = timedelta(days=settings.refresh_token_expire_days)

    expire = datetime.utcnow() + expires_delta
    to_encode = {
        "sub": user_id,
        "exp": expire,
        "type": "refresh",
    }
    token = jwt.encode(
        to_encode,
        settings.jwt_secret_key,
        algorithm=settings.jwt_algorithm,
    )
    return token, expire


def decode_token(token: str) -> dict:
    """Decode and validate a JWT token.

    Args:
        token: JWT token string.

    Returns:
        Decoded token payload.

    Raises:
        TokenError: If token is invalid or expired.
    """
    try:
        payload = jwt.decode(
            token,
            settings.jwt_secret_key,
            algorithms=[settings.jwt_algorithm],
        )
        return payload
    except JWTError as e:
        raise TokenError(f"Invalid token: {e}") from e


def get_user_id_from_token(token: str) -> str:
    """Extract user ID from a token.

    Args:
        token: JWT token string.

    Returns:
        User UUID from the token.

    Raises:
        TokenError: If token is invalid or doesn't contain user ID.
    """
    payload = decode_token(token)
    user_id = payload.get("sub")
    if not user_id:
        raise TokenError("Token does not contain user ID")
    return user_id


def verify_token_type(token: str, expected_type: str) -> bool:
    """Verify that a token is of the expected type.

    Args:
        token: JWT token string.
        expected_type: Expected token type ("access" or "refresh").

    Returns:
        True if token type matches, False otherwise.
    """
    try:
        payload = decode_token(token)
        return payload.get("type") == expected_type
    except TokenError:
        return False

File: frontend/src/lib/jwt.ts
import { SignJWT, jwtVerify, JWTPayload } from 'jose';

// Get secret as Uint8Array for jose library
const getSecret = () => new TextEncoder().encode(process.env.JWT_SECRET_KEY);

const ACCESS_TOKEN_EXPIRE_MINUTES = parseInt(
  process.env.ACCESS_TOKEN_EXPIRE_MINUTES || '15'
);
const REFRESH_TOKEN_EXPIRE_DAYS = parseInt(
  process.env.REFRESH_TOKEN_EXPIRE_DAYS || '7'
);
const REFRESH_TOKEN_EXPIRE_DAYS_REMEMBER = parseInt(
  process.env.REFRESH_TOKEN_EXPIRE_DAYS_REMEMBER || '30'
);

export class TokenError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'TokenError';
  }
}

export interface TokenPayload extends JWTPayload {
  sub: string;
  type: 'access' | 'refresh';
}

/**
 * Create a new access token
 */
export async function createAccessToken(
  userId: string,
  expiresInMinutes: number = ACCESS_TOKEN_EXPIRE_MINUTES
): Promise<string> {
  const token = await new SignJWT({ sub: userId, type: 'access' })
    .setProtectedHeader({ alg: 'HS256' })
    .setExpirationTime(`${expiresInMinutes}m`)
    .setIssuedAt()
    .sign(getSecret());

  return token;
}

/**
 * Create a new refresh token
 */
export async function createRefreshToken(
  userId: string,
  rememberMe: boolean = false
): Promise<{ token: string; expiresAt: Date }> {
  const days = rememberMe
    ? REFRESH_TOKEN_EXPIRE_DAYS_REMEMBER
    : REFRESH_TOKEN_EXPIRE_DAYS;

  const expiresAt = new Date(Date.now() + days * 24 * 60 * 60 * 1000);

  const token = await new SignJWT({ sub: userId, type: 'refresh' })
    .setProtectedHeader({ alg: 'HS256' })
    .setExpirationTime(`${days}d`)
    .setIssuedAt()
    .sign(getSecret());

  return { token, expiresAt };
}

/**
 * Decode and validate a JWT token
 */
export async function decodeToken(token: string): Promise<TokenPayload> {
  try {
    const { payload } = await jwtVerify(token, getSecret());
    return payload as TokenPayload;
  } catch (error) {
    throw new TokenError(`Invalid token: ${error}`);
  }
}

/**
 * Extract user ID from a token
 */
export async function getUserIdFromToken(token: string): Promise<string> {
  const payload = await decodeToken(token);
  const userId = payload.sub;

  if (!userId) {
    throw new TokenError('Token does not contain user ID');
  }

  return userId;
}

/**
 * Verify that a token is of the expected type
 */
export async function verifyTokenType(
  token: string,
  expectedType: 'access' | 'refresh'
): Promise<boolean> {
  try {
    const payload = await decodeToken(token);
    return payload.type === expectedType;
  } catch {
    return false;
  }
}

/**
 * Cookie options for access token
 */
export const accessTokenCookieOptions = {
  httpOnly: true,
  secure: process.env.NODE_ENV === 'production',
  sameSite: 'lax' as const,
  maxAge: ACCESS_TOKEN_EXPIRE_MINUTES * 60,
  path: '/',
};

/**
 * Cookie options for refresh token
 */
export function getRefreshTokenCookieOptions(rememberMe: boolean = false) {
  const days = rememberMe
    ? REFRESH_TOKEN_EXPIRE_DAYS_REMEMBER
    : REFRESH_TOKEN_EXPIRE_DAYS;

  return {
    httpOnly: true,
    secure: process.env.NODE_ENV === 'production',
    sameSite: 'lax' as const,
    maxAge: days * 24 * 60 * 60,
    path: '/',
  };
}
